File: c2-project/Solution52.cpp
//
// Created by 陈高杰 on 2020/2/16.
//
#include <iostream>
#include <stdlib.h>
#include <string>
#include <vector>
#include <stack>
#include <queue>
#include <math.h>
#include <sstream>
#include <map>
using namespace std;


struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};

class Solution {
public:
    ListNode* FindFirstCommonNode( ListNode* pHead1, ListNode* pHead2) {
        int length1 = GetListLength(pHead1);
        int length2 = GetListLength(pHead2);
        int difidx = length1 - length2;
        ListNode* pLong = pHead1;
        ListNode* pShort = pHead2;
        if(length1<length2){
            ListNode* pLong = pHead2;
            ListNode* pShort = pHead1;
            difidx = length2 - length1;
        }

        for (int i = 0; i < difidx; ++i) {
            pLong = pLong->next;
        }
        while(pLong!= nullptr && pShort!= nullptr && pLong!=pShort){
            pLong = pLong->next;
            pShort = pShort->next;
        }
        ListNode* pCommon = pLong;
        return pCommon;
    }

    int GetListLength(ListNode* pHead1) {
        int length = 0;
        ListNode* pNode = pHead1;
        while(pNode !=nullptr){
            length++;
            pNode = pNode->next;
        }
        return length;
    }
};

//int main(){
//    ListNode ListNode1 = ListNode(1);
//    ListNode ListNode2 = ListNode(2);
//    ListNode ListNode3 = ListNode(3);
//    ListNode ListNode4 = ListNode(4);
//    ListNode ListNode5 = ListNode(5);
//    ListNode ListNode6 = ListNode(6);
//    ListNode ListNode7 = ListNode(7);
//
//    ListNode1.next = &ListNode2;
//    ListNode2.next = &ListNode3;
//    ListNode3.next = &ListNode6;
//    ListNode6.next = &ListNode7;
//
//    ListNode4.next = &ListNode5;
//    ListNode5.next = &ListNode6;
//
//    Solution solution;
//    ListNode* x = solution.FindFirstCommonNode(&ListNode1,&ListNode4);
//    cout<< x->val<<endl;
//    system("PAUSE");
//    return 0;
//}

File: py-project/Solution52.py
# -*- coding:utf-8 -*-
import functools
import sys

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def FindFirstCommonNode(self, pHead1, pHead2):
        # write code here
        length1 = self.GetListLength(pHead1)
        length2 = self.GetListLength(pHead2)
        diffidx = length1 - length2
        pLong = pHead1
        pShort = pHead2

        if diffidx <0:
            pLong = pHead2
            pShort = pHead1
            diffidx = length2 - length1

        for i in range(diffidx):
            pLong = pLong.next

        while pLong != None and pShort != None and pLong != pShort:
            pLong = pLong.next
            pShort = pShort.next

        pNode = pLong
        return pNode

    def GetListLength(self,pNode):
        length = 0
        while pNode != None:
            pNode = pNode.next
            length +=1
        return length


if __name__ == '__main__':
    ListNode1 = ListNode(1)
    ListNode2 = ListNode(2)
    ListNode3 = ListNode(3)
    ListNode4 = ListNode(4)
    ListNode5 = ListNode(5)
    ListNode6 = ListNode(6)
    ListNode7 = ListNode(7)

    ListNode1.next = ListNode2
    ListNode2.next = ListNode3
    ListNode3.next = ListNode6
    ListNode4.next = ListNode5
    ListNode5.next = ListNode6
    ListNode6.next = ListNode7

    # print(len(l))
    solution1 = Solution()
    a = solution1.FindFirstCommonNode(ListNode1,ListNode4)
    print(a.val)